File: 백준/Gold/7569. 토마토/토마토.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {
    static int[][][] arr = new int[105][105][105];
    static int[][][] dist = new int[105][105][105];
    static int[] dx = {1,0,-1,0,0,0};
    static int[] dy = {0,1,0,-1,0,0};
    static int[] dz = {0,0,0,0,1,-1};
    static class Triple{
        int x,y,z;
        Triple(int x, int y, int z){
            this.x = x;
            this.y = y;
            this.z = z;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] str = br.readLine().split(" ");
        int m = Integer.parseInt(str[0]);
        int n = Integer.parseInt(str[1]);
        int h = Integer.parseInt(str[2]);

        Queue<Triple> queue = new LinkedList<>();
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < n; j++) {
                str = br.readLine().split(" ");
                for (int k = 0; k < m; k++) {
                    arr[k][j][i] = Integer.parseInt(str[k]);
                    if (arr[k][j][i] == 1) queue.add(new Triple(k, j, i));
                    if (arr[k][j][i] == 0) dist[k][j][i] = -1;
                }
            }
        }



        while(!queue.isEmpty()){
            Triple cur = queue.poll();
            for (int i = 0; i < 6; i++) {
                int nx = cur.x + dx[i];
                int ny = cur.y + dy[i];
                int nz = cur.z + dz[i];
                if (nx < 0 || ny < 0 || nz < 0 || nx >= m || ny >= n || nz >= h) continue;
                if (dist[nx][ny][nz] >= 0) continue;
                dist[nx][ny][nz] = dist[cur.x][cur.y][cur.z] + 1;
                queue.add(new Triple(nx, ny, nz));
            }
        }

        boolean flag = true;
        int cnt = -1;
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < m; k++) {
                    if (dist[k][j][i] == -1){
                        flag = false;
                        break;
                    }
                    cnt = Math.max(cnt, dist[k][j][i]);
                }
                if(!flag) break;
            }
            if(!flag) break;
        }

        if(!flag) System.out.println(-1);
        else System.out.println(cnt);
    }
}

File: 백준/Gold/7569. 토마토/토마토.py
import sys, copy
from collections import *

input = sys.stdin.readline

m, n, h = map(int, input().split())  # 가로, 세로, 높이

box = [[list(map(int, input().split())) for _ in range(n)] for _ in range(h)]
dx = [1, 0, -1, 0, 0, 0]
dy = [0, 1, 0, -1, 0, 0]
dz = [0, 0, 0, 0, 1, -1]
q = deque()
dist = [[[0] * m for _ in range(n)] for _ in range(h)]
for j in range(m):
    for i in range(n):
        for k in range(h):
            if box[k][i][j] == 1:
                q.append((i, j, k, 0))
            if box[k][i][j] == 0:
                dist[k][i][j] = -1
while q:
    x, y, z, cnt = q.popleft()
    for i in range(6):
        nx = x + dx[i]
        ny = y + dy[i]
        nz = z + dz[i]
        if nx < 0 or ny < 0 or nz < 0 or nx >= n or ny >= m or nz >= h:
            continue
        if dist[nz][nx][ny] >= 0:
            continue
        dist[nz][nx][ny] = dist[z][x][y] + 1
        q.append((nx, ny, nz, cnt + 1))

ans = 0
for j in range(m):
    for i in range(n):
        for k in range(h):
            if dist[k][i][j] == -1:
                print(-1)
                exit()
            ans = max(ans, dist[k][i][j])
print(ans)
